image-prune

```diff
@@ -152,34 +140,74 @@ def prune_images(force, dryrun, quiet=False, open_pull_requests=True, offline=Fa
             target = os.readlink(entry.path)
             targets.add(target)
 
-    expiry_threshold = now - IMAGE_EXPIRE * 86400
-    for entry in os.scandir(get_images_data_dir()):
-        if not force and (enough_disk_space() and entry.stat(follow_symlinks=False).st_mtime > expiry_threshold):
-            continue
-        if entry.is_file() and \
-                any(entry.name.endswith(ext) for ext in ['.qcow2', '.partial']) and \
-                entry.name not in targets:
-            if not quiet or dryrun:
-                sys.stderr.write("Pruning {0}\n".format(entry.name))
+    return targets
+
+
+class ImageCache:
+    def prune(self, keepers, force=False, dryrun=False, quiet=False):
+        expiry_threshold = time.time() - IMAGE_EXPIRE * 86400
+
+        for image, mtime in self.list_files():
+            if not any(image.endswith(ext) for ext in ['.qcow2', '.partial']):
+                debug(f'Skipping file {image} with unknown extension')
+                continue
+
+            if image in keepers:
+                debug(f'Skipping image {image} which is in the keepers list')
+                continue
+
+            if not force and self.enough_space() and mtime > expiry_threshold:
+                debug(f'Skipping image {image} which is new enough (and not forced or low on space)')
+                continue
+
+            if not quiet:
+                print(f"Pruning {image}", file=sys.stderr)
+
             if not dryrun:
-                os.unlink(entry.path)
+                self.delete_image(image)
+
+        self.cleanup(quiet, dryrun)
+
+
+class LocalImageDirectory(ImageCache):
+    def __init__(self, directory=None):
+        self.directory = directory or get_images_data_dir()
+        self.space_threshold = float(os.environ.get("PRUNE_THRESHOLD_G", 15))
```

Comment:
btw, if it helps to simplify things for S3: Feel free to remove the entire "check free disk space" and that env var. These days, the solution for "we don't have enough disk space for our purposes" is "turn the diskspace knob".

Reply:
I actually want to implement the disk space thing for S3: it can help us make sure the stores stay under a given fixed size.